asyncio_run_in_process/run_in_process.py

```diff
@@ -0,0 +1,233 @@
+    logger.debug("waiting for process %s finish", proc)
+
+    await sub_proc.wait()
+
+    proc.returncode = sub_proc.returncode
+    logger.debug("process %s finished: returncode=%d", proc, proc.returncode)
+
+
+async def _relay_signals(
+    proc: Process[Any],
+    queue: 'asyncio.Queue[int]',
+) -> None:
+    if proc.state.is_before(State.EXECUTING):
+        # If the process has not reached the state where the child process
+        # can properly handle the signal, give it a moment to reach the
+        # `EXECUTING` stage.
+        await proc.wait_for_state(State.EXECUTING)
+
+    while True:
+        signum = await queue.get()
+
+        logger.debug("relaying signal %s to child process %s", signum, proc)
+        proc.send_signal(signum)
+
+
+async def _monitor_state(
+    proc: Process[TReturn], from_child: BinaryIO, loop: asyncio.AbstractEventLoop,
+) -> None:
+    for expected_state in State:
+        if proc.state is not expected_state:
+            raise InvalidState(
+                f"Process in state {proc.state} but expected state {expected_state}"
+            )
+
+        child_state_as_byte = await loop.run_in_executor(None, read_exactly, from_child, 1)
+
+        try:
+            child_state = State(child_state_as_byte)
+        except TypeError:
+            raise InvalidState(f"Child sent state: {child_state_as_byte.hex()}")
+
+        if not proc.state.is_next(child_state):
+            if proc.state is State.FINISHED:
+                # This case covers when the process is killed with a SIGKILL
+                # and jumps directly to the finished state.
+                return
+            raise InvalidState(
+                f"Invalid state transition: {proc.state} -> {child_state}"
+            )
+
+        if child_state is State.FINISHED:
+            # For the FINISHED state we delay updating the state until we also
+            # have a return value.
+            break
+
+        proc.state = child_state
+        logger.debug(
+            "Updated process %s state %s -> %s",
+            proc,
+            expected_state.name,
+            child_state.name,
+        )
+
+    # This is mostly a sanity check but it ensures that the loop variable is
+    # what we expect it to be before starting to collect the result the stream.
+    if child_state is not State.FINISHED:
+        raise InvalidState(f"Invalid final state: {proc.state}")
+
+    result = await loop.run_in_executor(None, receive_pickled_value, from_child)
+
+    await proc.wait_returncode()
+
+    if proc.returncode == 0:
+        proc.return_value = result
+    else:
+        proc.error = result
+
+    proc.state = child_state
+    logger.debug(
+        "Updated process %s state %s -> %s",
+        proc,
+        expected_state.name,
+        child_state.name,
+    )
+
+
+RELAY_SIGNALS = (signal.SIGINT, signal.SIGTERM, signal.SIGHUP)
+
+
+@asynccontextmanager
+async def open_in_process(
+    async_fn: Callable[..., TReturn],
+    *args: Any,
```

Comment:
Are these compatible with the standard args for `Popen`  such as `cwd`?

Reply:
No, these are the `*args` that will be passed into `async_fn` when called in the new process.